File: packages/java/ygo74-agent-runtime/src/main/java/com/ygo74/agentruntime/middleware/MiddlewareRegistry.java
package com.ygo74.agentruntime.middleware;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public final class MiddlewareRegistry {
    public record Registered(String id, Middleware middleware, int order) {
    }

    private final List<Registered> items = new ArrayList<>();

    public void register(String id, Middleware middleware, int order) {
        items.add(new Registered(id, middleware, order));
    }

    public List<Registered> ordered() {
        return items.stream().sorted(Comparator.comparingInt(Registered::order)).toList();
    }
}

File: packages/dotnet/Ygo74.AgentRuntime/Middleware/MiddlewareRegistry.cs
namespace Ygo74.AgentRuntime.Middleware;

public sealed record RegisteredMiddleware(string Id, IMiddleware Middleware, int Order);

public sealed class MiddlewareRegistry
{
    private readonly List<RegisteredMiddleware> _items = new();

    public void Register(string id, IMiddleware middleware, int order)
    {
        _items.Add(new RegisteredMiddleware(id, middleware, order));
    }

    public IReadOnlyList<RegisteredMiddleware> Ordered() => _items.OrderBy(x => x.Order).ToList();
}
